src/crypto/ecdh.rs

```diff
@@ -252,25 +250,41 @@ pub fn encrypt<R: CryptoRng + Rng>(
     // Perform key derivation
     let z = kdf(hash, shared_secret.as_bytes(), alg_sym.key_size(), &param)?;
 
-    // PKCS5 padding (PKCS5 is PKCS7 with a blocksize of 8)
+    // PKCS5-style padding, with a blocksize of 8.
+    // However, the padding may exceed the length one block, to obfuscate key size.
     let len = plain.len();
-    let mut plain_padded = plain.to_vec();
-    plain_padded.resize(len + 8, 0);
 
-    let plain_padded_ref = {
-        let pos = len;
-        let block_size = 8;
-        let bs = block_size * (pos / block_size);
-        if plain_padded.len() < bs || plain_padded.len() - bs < block_size {
-            bail!("unable to pad");
-        }
-        let buf = GenericArray::<u8, U8>::from_mut_slice(&mut plain_padded[bs..bs + block_size]);
-        Pkcs7::pad(buf, pos - bs);
-        &plain_padded[..bs + block_size]
+    /// Our default target padded length (based on the size of a padded AES256 key).
+    /// This value should be increased if we support symmetric keys that are longer than AES256.
+    const PAD_DEFAULT_TARGET: usize = 40;
+
+    // The padded message length (must be a multiple of the block size)
+    let padded_len = if len < PAD_DEFAULT_TARGET {
+        // Normally, we just pad to the default target size ...
+        PAD_DEFAULT_TARGET
+    } else {
+        // ... but if `plain` isn't shorter than our target size, we pad to the next full block
+        let remainder = len % 8; // e.g. 3 for len==19
+
+        len + 8 - remainder // (e.g. "8 + 8 - 0 => 16", or "19 + 8 - 3 => 24")
     };
+    ensure!(padded_len % 8 == 0, "Unexpected padded_len {}", padded_len);
+
+    // The value we'll use for padding (must be between 1 and 255)
+    let padding = padded_len - len;
+    ensure!(
+        padding > 0 && padding < 256,
```

Comment:
you could use `u8::try_from(padding)` for the size check instead

Reply:
i also want >0, so `u8::try_from` is not exactly shorter - but i agree that it's more expressive. changed.